sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSource.java

```diff
@@ -402,11 +401,11 @@ private boolean readCustomLine() throws IOException {
       delimiterFinder.reset();
 
       while (true) {
-        int startPosn = bufferPosn;
-        while (bufferPosn >= bufferLength) {
-          startPosn = bufferPosn = 0;
+        if (bufferPosn >= bufferLength) {
           byteBuffer.clear();
-          bufferLength = inChannel.read(byteBuffer);
+          do {
+            bufferLength = inChannel.read(byteBuffer);
+          } while (bufferLength == 0);
 
           if (bufferLength < 0) {
             eof = true;
```

Comment:
need break here now

Reply:
It does `break` if it's EOF. Then, it quite the big loop and return the current value. When it reads some data, it goes to find a delimiter inside the `buffer`.

```
      while (true) {
        if (bufferPosn >= bufferLength) { <-- changed from `while` to `if`
          byteBuffer.clear();
          do {
            bufferLength = inChannel.read(byteBuffer); <-- looping until read something or EOF.
          } while (bufferLength == 0);

          if (bufferLength < 0) {
            eof = true;

            if (str.size() == 0) {
              return false;
            }

            // Not ending with a delimiter.
            currentValue = str.toString(StandardCharsets.UTF_8.name());
            break;  <-- break here
          }
        }

        ... (continue to find the delimiter in the buffer)...
     
```